lib/src/main/java/com/knuddels/jtokkit/TokenEncoderLarge.java

```diff
@@ -33,46 +33,49 @@ static int calculateTokensLarge(TokenEncoder tokenEncoder, int maxTokenCount, bo
-            RankNode nextNextNode = nextNode != null ? nextNode.next : null;
-            RankNode nextNextNextNode = nextNextNode != null ? nextNextNode.next : null;
-
-            if (previousNode != null) {
-                int newRank = tokenEncoder.encode(match, previousNode.index, nextNextNode != null ? nextNextNode.index : Integer.MAX_VALUE);
-                if ((newRank == MAX_RANK) != (previousNode.rank == MAX_RANK)) {
-                    validRanks -= (newRank == MAX_RANK) ? 1 : -1;
+            TreeMap<Integer, RankNode> minNodes = rankMap.firstEntry().getValue();
+            for (int i = 0; i < minNodes.size(); i++) {
+                RankNode minNode = minNodes.firstEntry().getValue();
+                assert minNode.rank != MAX_RANK;
+
+                RankNode previousNode = minNode.prev;
+                RankNode nextNode = minNode.next;
+                RankNode nextNextNode = nextNode != null ? nextNode.next : null;
+                RankNode nextNextNextNode = nextNextNode != null ? nextNextNode.next : null;
+
+                if (previousNode != null) {
+                    int newRank = tokenEncoder.encode(match, previousNode.index, nextNextNode != null ? nextNextNode.index : Integer.MAX_VALUE);
+                    if ((newRank == MAX_RANK) != (previousNode.rank == MAX_RANK)) {
+                        validRanks -= (newRank == MAX_RANK) ? 1 : -1;
+                    }
+                    removeNode(rankMap.get(previousNode.rank), rankMap, previousNode);
+                    previousNode.rank = newRank;
+                    rankMap.computeIfAbsent(newRank, k -> new TreeMap<>()).put(previousNode.index, previousNode);
                 }
-                removeNode(rankMap, previousNode);
-                previousNode.rank = newRank;
-                rankMap.computeIfAbsent(newRank, k -> new TreeMap<>()).put(previousNode.index, previousNode);
-            }
 
-            int newRank = tokenEncoder.encode(match, minNode.index, nextNextNextNode != null ? nextNextNextNode.index : Integer.MAX_VALUE);
-            if ((newRank == MAX_RANK) != (minNode.rank == MAX_RANK)) {
-                validRanks--;
-            }
-            removeNode(rankMap, minNode);
-            minNode.rank = newRank;
-            rankMap.computeIfAbsent(newRank, k -> new TreeMap<>()).put(minNode.index, minNode);
-
-            minNode.next = nextNextNode;
-            if (nextNode != null) {
-                if (nextNextNode != null) {
-                    nextNextNode.prev = minNode;
-                }
-                if (nextNode.rank != MAX_RANK) {
+                int newRank = tokenEncoder.encode(match, minNode.index, nextNextNextNode != null ? nextNextNextNode.index : Integer.MAX_VALUE);
+                if ((newRank == MAX_RANK) != (minNode.rank == MAX_RANK)) {
                     validRanks--;
                 }
-                removeNode(rankMap, nextNode);
-            }
+                removeNode(minNodes, rankMap, minNode);
```

Comment:
minor optimization - we could probably do a few more tiny ones here, but this is a worst-case handler anyway